### DISCUSSION TITLE
Android 13 per-app language support

### DISCUSSION THREAD

**Discussion by kryslt** [apache/cordova#458] [General]
Hello, this is a new and VERY USEFUL feature in Android 13. Do you have any information about this feature in cordova (or cordova-android)?

https://developer.android.com/guide/topics/resources/app-languages#auto-localeconfig

Thanks

**Comment by breautek** [comment]
Given that it requires AGP 8.1 it will require https://github.com/apache/cordova-android/pull/1678

But most resources in cordova apps are web resources and can't really take advantage of the native locale system. So this will only really be useful if you have plugins that builds native views and displays text to the user at the native view layer (outside of the webview). A plugin can probably be implemented to expose the build configurations required to setup the auto locale config.

Note that The webview does have an [INTL](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) API to handle formatting of text and numbers, and most JS frameworks incorporates some kind of localisation string table system. If you're trying to localize your webview content, a pure webview solution will likely be far more efficient.

**Reply by kryslt** [reply]
I use automatically generated JS arrays with different localization strings that load according to webview locale. The per-app language selection works even for this, i.e. the webview indicates correct locale as selected by the system per-app picker. Also see my answer below.

**Comment by kryslt** [comment]
I reply to myself. 
For some parts of this new feature Cordova has enough options now. 
I've just implemented the manual setup described here https://developer.android.com/guide/topics/resources/app-languages#use-localeconfig and it works. 

Two steps are necessary:
1. Create the file in a separate source directory (e.g in _src/locales_config.xml_) and copy it to _platforms/android/app/src/main/res/xml_ before compiling. Use hooks in your config.xml file:

`<hook type="before_compile" src="hook-android-before-compile.js" />`

hook-android-before-compile.js: 

```
var path = require('path');

function copyAndroidFiles() {
  const fs = require('fs-extra');
  fs.copySync("src/locales_config.xml", "platforms/android/app/src/main/res/xml/locales_config.xml");
}

module.exports = function(context) {
  copyAndroidFiles();
};

```

2. Add option to Android manifest in your config.xml file:
```
 <platform name="android">
...
    <edit-config file="AndroidManifest.xml" target="/manifest/application" mode="merge">
      <application android:localeConfig="@xml/locales_config" />
...
    </edit-config>
...

```

Build and run your Cordova app on your phone. The app appears in the App languages menu. You can now select language only for your app. The app (or only the internal webview, I don't know exactly) restarts with a new language code. 

This Android 13 feature is really VERY USEFUL because there is one general method to select language for any app (provided it is properly configured). No need for other extra JS code in your Cordova apps. I've used language files in my apps so far but never (luckily!) tried to implement own in-app language picker and corresponding code.

I'm not sure how to implement the automatic generation of the locales_config.xml described here https://developer.android.com/guide/topics/resources/app-languages#auto-localeconfig but for me the manual process is simple.

I don't need it, but I'm not sure how to implement an in-app language picker as described here https://developer.android.com/guide/topics/resources/app-languages#api-implementation, either.
I believe here Cordova needs to support the setApplicationLocales() and getApplicationLocales() methods, i.e. make them available for the webview code. Maybe through this plugin https://cordova.apache.org/docs/en/6.x/reference/cordova-plugin-globalization/.
